### PR TITLE
Add missing boundary checks to cunn_SoftMaxForward

### PR DESCRIPTION
**Pull request by xinyazhang**
This fixes OOB memory access for following code
```python
import torch
qk = torch.randn((1024,587), dtype=torch.float64, device='cuda')
smqk = torch.softmax(qk, dim=-1)
```

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd

### CODE REVIEW

**Comment by xinyazhang** [comment]
The segfault can be reproduced on ROCM platform with the following env vars
```
PYTORCH_NO_HIP_MEMORY_CACHING=1 HSA_SVM_GUARD_PAGES=1 HSA_DISABLE_FRAGMENT_ALLOCATOR=1 AMD_SERIALIZE_KERNEL=3 
```
, which add guard pages to detect OOB memory access.

**Comment by xinyazhang** [comment]
@jithunnair-amd  @pruthvistony  @jataylo @jeffdaily 
I'm speculating our large SDPA error is partially due to the OOB access in the softmax operator, which reads extra data when the input is irregular.

**Review by jeffdaily — APPROVED** [review]

**Review by eqy — COMMENTED** [review]
If you have a reproducer, could you add a test case for this change please?

**Review by malfet — APPROVED** [review]
@xinyazhang please add some sort of a test to the PR description, but otherwise looks good to me (we can not have unittest unfortunately, because at least on CUDA something like that would not be detectable without compute sanitizer)

**Comment by xinyazhang** [comment]
> If you have a reproducer, could you add a test case for this change please?

@eqy I have a testing case in mind, but where should I add it? (or according to @malfet we cannot add it)

**Comment by eqy** [comment]
> @xinyazhang please add some sort of a test to the PR description, but otherwise looks good to me (we can not have unittest unfortunately, because at least on CUDA something like that would not be detectable without compute sanitizer)

Would we get some signal with `torch.cuda.empty_cache()` around the test?

**Comment by malfet** [comment]
> > @xinyazhang please add some sort of a test to the PR description, but otherwise looks good to me (we can not have unittest unfortunately, because at least on CUDA something like that would not be detectable without compute sanitizer)
> 
> Would we get some signal with `torch.cuda.empty_cache()` around the test?

Nope, as GPU maps memory into its address space in a pretty large chunks, see
```
$ PYTORCH_NO_CUDA_MEMORY_CACHING=1 python -c "import torch;qk = torch.randn((1024,587), dtype=torch.float64, device='cuda'); print(torch.softmax(qk, dim=-1))"; echo $?
0
```
vs launching the same with compute sanitizer, that will catch an error

**Comment by xinyazhang** [comment]
For documentation @malfet @eqy @jeffdaily @jithunnair-amd  here is the independent unit test code.
```
#!/usr/bin/env python

import torch
import pytest

@pytest.mark.parametrize("dtype",
                         [torch.float16, torch.bfloat16, torch.float32, torch.float64],
                         ids=['fp16', 'bf16', 'fp32', 'fp64'])
def test_softmax_oob_access(dtype):
    qk_with_margin = torch.randn((1024+1, 587), dtype=dtype, device="cuda");
    qk_with_margin[-1].fill_(float('nan'))
    qk = qk_with_margin[:-1, :]
    smqk = torch.softmax(qk, dim=-1)
    assert not torch.isnan(smqk).any(), 'NaN indicates OOB memory access'
```

It only fails on float64 apparently, (but our SDPA's UTs suffers from it and has to use CPU implementation on certain cases)

**Comment by malfet** [comment]
@xinyazhang that's smart and indeed reliable, please do not hesitate to propose a PR that adds this unites
